handler/category-dimension-import.go

```diff
@@ -186,6 +143,93 @@ func (h *CategoryDimensionImport) Handle(ctx context.Context, e *event.CategoryD
+
+	// process full batches
+	for i := 0; i < numFullChunks; i++ {
+		batchOffset := i * h.cfg.BatchSizeLimit
+		optionsBatch := make([]*dataset.OptionPost, h.cfg.BatchSizeLimit)
+		for j := 0; j < h.cfg.BatchSizeLimit; j++ {
+			optionsBatch[j] = &dataset.OptionPost{
+				Name:     variable.Name,
+				CodeList: variable.Name,                 // TODO can we assume this?
+				Code:     variable.Codes[batchOffset+j], // TODO can we assume this?
+				Option:   variable.Codes[batchOffset+j],
+				Label:    variable.Labels[batchOffset+j],
+			}
+		}
+		eTag, err = h.PatchInstanceDimensionsWithRetires(ctx, e, optionsBatch, eTag, 0)
+		if err != nil {
+			return "", err
+		}
+	}
+
+	// process any remaining options in a last smaller batch
+	if remainingSize > 0 {
+		batchOffset := numFullChunks * h.cfg.BatchSizeLimit
+		optionsBatch := make([]*dataset.OptionPost, remainingSize)
+		for j := 0; j < remainingSize; j++ {
+			optionsBatch[j] = &dataset.OptionPost{
+				Name:     variable.Name,
+				CodeList: variable.Name,                 // TODO can we assume this?
+				Code:     variable.Codes[batchOffset+j], // TODO can we assume this?
+				Option:   variable.Codes[batchOffset+j],
+				Label:    variable.Labels[batchOffset+j],
+			}
+		}
+		eTag, err = h.PatchInstanceDimensionsWithRetires(ctx, e, optionsBatch, eTag, 0)
+		if err != nil {
+			return "", err
+		}
+	}
+
+	return eTag, nil
+}
+
+// PatchInstanceDimensionsWithRetires sends a patch request to dataset API with the provided options.
+// If the eTag value changes, validate the instance state again and retry only if it is still in 'completed' state.
+// We will do up to MaxConflictRetries retires, if all of them fail, set the instance and import job to failed state
+func (h *CategoryDimensionImport) PatchInstanceDimensionsWithRetires(ctx context.Context, e *event.CategoryDimensionImport, options []*dataset.OptionPost, eTag string, attempt int) (newETag string, err error) {
+	eTag, err = h.datasets.PatchInstanceDimensions(ctx, h.cfg.ServiceAuthToken, e.InstanceID, options, eTag)
+	if err != nil {
+		switch errPost := err.(type) {
+		// ErrInvalidDatasetAPIResponse covers the case where the dataset API responded with an unexpected Status Code.
+		// If the status code was 409 Conflict, then it means that the instance changed since the last call.
+		case *dataset.ErrInvalidDatasetAPIResponse:
+			if errPost.Code() == http.StatusConflict {
+				// check if we have already attemtped to post the instance more than MaxConflictRetries times
+				if attempt >= MaxConflictRetries {
+					return "", h.setImportToFailed(ctx, fmt.Errorf("aborting import process after %d retries resulting in conflict on post dimension", MaxConflictRetries), e)
+				}
+
+				// sleep an exponential random time before retrying
+				SleepRandom(attempt)
+
+				// check that the instance is still in 'completed' state
+				_, eTag, err = h.getCompletedInstance(ctx, e, headers.IfMatchAnyETag)
+				if err != nil {
+					return "", err
```

Comment:
No context for error?

Reply:
I thought that the error in `getCompletedInstance` was enough, but I will add a wrap here as well